<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#UAS-Lab-Data-Mining" data-toc-modified-id="UAS-Lab-Data-Mining-1">UAS Lab Data Mining</a></span><ul class="toc-item"><li><span><a href="#Import-Library-&amp;-Dataset" data-toc-modified-id="Import-Library-&amp;-Dataset-1.1">Import Library &amp; Dataset</a></span></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-1.2">Data Preprocessing</a></span><ul class="toc-item"><li><span><a href="#Ganti-?-dengan-NaN" data-toc-modified-id="Ganti-?-dengan-NaN-1.2.1">Ganti ? dengan NaN</a></span></li><li><span><a href="#Drop-instance-dengan-NaN" data-toc-modified-id="Drop-instance-dengan-NaN-1.2.2">Drop instance dengan NaN</a></span></li><li><span><a href="#Encode-Categorical-Data" data-toc-modified-id="Encode-Categorical-Data-1.2.3">Encode Categorical Data</a></span></li><li><span><a href="#Normalisasi-Dataset" data-toc-modified-id="Normalisasi-Dataset-1.2.4">Normalisasi Dataset</a></span></li><li><span><a href="#Simpan-Final-Dataset" data-toc-modified-id="Simpan-Final-Dataset-1.2.5">Simpan Final Dataset</a></span></li></ul></li><li><span><a href="#Membangun-Model" data-toc-modified-id="Membangun-Model-1.3">Membangun Model</a></span><ul class="toc-item"><li><span><a href="#Train-Test-Split" data-toc-modified-id="Train-Test-Split-1.3.1">Train Test Split</a></span><ul class="toc-item"><li><span><a href="#KNN-dan-GridSearch" data-toc-modified-id="KNN-dan-GridSearch-1.3.1.1">KNN dan GridSearch</a></span></li></ul></li><li><span><a href="#Using-Data-Testing" data-toc-modified-id="Using-Data-Testing-1.3.2">Using Data Testing</a></span></li></ul></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-1.4">Kesimpulan</a></span></li><li><span><a href="#Referensi" data-toc-modified-id="Referensi-1.5">Referensi</a></span></li></ul></li></ul></div>

# UAS Lab Data Mining

## Import Library & Dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
train_path = '../input/adult.data'
test_path = '../input/adult.test'

In [3]:
# informasi nama feature
colnames = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 
            'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 
            'salary']

# load dataset
train = pd.read_csv(train_path, names=colnames, skipinitialspace=True)
test = pd.read_csv(test_path, names=colnames, skipinitialspace=True)

In [4]:
train.head(5)

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [5]:
test.head(5)

age  workclass  fnlwgt     education  education-num      marital-status  \
0   25    Private  226802          11th              7       Never-married   
1   38    Private   89814       HS-grad              9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm             12  Married-civ-spouse   
3   44    Private  160323  Some-college             10  Married-civ-spouse   
4   18          ?  103497  Some-college             10       Never-married   

          occupation relationship   race     sex  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country  salary  
0              40  United-States  <=50K.  
1              50  United-States  <=50K.  
2              40  United-States   >50K.  
3              40  United-States   >50K.  
4              30  United-States  <=50K.

In [6]:
print(
    f'''
    Dimensi data train: {train.shape}
    Dimensi data test: {test.shape}''')


    Dimensi data train: (32561, 15)
    Dimensi data test: (16281, 15)


## Data Preprocessing

### Ganti ? dengan NaN

In [7]:
# ganti ? dengan NaN
train.replace({'?':np.nan}, 
              inplace=True)

# ganti ? dengan NaN
test.replace({'?':np.nan}, 
             inplace=True)

# ganti nilai karena feature salary diakhiri dengan .
test.replace({'salary': {'<=50K.':'<=50K', '>50K.':'>50K'}}, 
             inplace=True)

### Drop instance dengan NaN

In [8]:
# cek jumlah NaN
print(
    f'''
    Jumlah NaN di data train: {train.isna().sum().sum()}
    Jumlah NaN di data test: {test.isna().sum().sum()}''')


    Jumlah NaN di data train: 4262
    Jumlah NaN di data test: 2203


In [9]:
# drop NaN value
train.dropna(axis='index', 
             inplace=True)

test.dropna(axis='index', 
            inplace=True)

In [10]:
print(
    f'''
    Dimensi data train: {train.shape}
    Dimensi data test: {test.shape}''')


    Dimensi data train: (30162, 15)
    Dimensi data test: (15060, 15)


Dimensi data train dan test sesuai dengan informasi yang diberikan di **adult.names**.

### Encode Categorical Data

Data categorical harus diencode terlebih dahulu. Saya menggunakan **OrdinalEncoder** dari *sklearn*.

In [11]:
# select categorical feature
categorical_columns = list( train.select_dtypes(include=['object']).columns )
categorical_columns

['workclass',
 'education',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native-country',
 'salary']

In [12]:
from sklearn.preprocessing import OrdinalEncoder

In [13]:
oe = OrdinalEncoder()

# encode categorical data di data train
train[categorical_columns] = oe.fit_transform(train[categorical_columns])

# encode categorical data di data test
test[categorical_columns] = oe.fit_transform(test[categorical_columns])

### Simpan Final Dataset

In [14]:
# simpan data training
train.to_csv('../output/adult-training.csv', 
             index=False)

# simpan data testing
test.to_csv('../output/adult-test.csv', 
            index=False)

# concate dataframe train dan test
complete = pd.concat(objs=[train, test], 
                     ignore_index=True)

# simpan complete-encoded
complete.to_csv('../output/adult-complete.csv', 
                index=False)

## Membangun Model

Diminta untuk membangun model dengan 2 cara yang berbeda yaitu menggunakan fungsi`train_test_split` dan menggunakan file `adult-testing.csv`.

### Train Test Split

In [15]:
# import dataset complete
complete = pd.read_csv('../output/adult-complete.csv')

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
# drop feature salary
X = complete.drop(columns=['fnlwgt', 'salary'])

# simpan feature salary 
y = complete.salary

In [18]:
# ukuran test size 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.2, 
                                                    random_state=42)

#### KNN dan GridSearch

In [19]:
from sklearn.neighbors import KNeighborsClassifier

Dari parameter tuning yang telah dilakukan, parameter terbaik yang menghasilkan score tertinggi yaitu:
- **n**: 23
- **metric**: manhattan
- **weights**: distance

Dari parameter diatas, model KNN akan dibangun kembali.

In [20]:
# bangun model
model = KNeighborsClassifier(n_neighbors=23, 
                             weights='distance', 
                             metric='manhattan')

# fit data training
model.fit(X_train, y_train)

# test model
y_pred = model.predict(X_test)

In [21]:
from sklearn.metrics import accuracy_score, classification_report

# print akurasi
print(f"Akurasi: {accuracy_score(y_test, y_pred):.4f}")

# print precision, recall & F-1 Score
print(classification_report(y_pred=y_pred, 
                            y_true=y_test))

Akurasi: 0.8547
              precision    recall  f1-score   support

         0.0       0.89      0.93      0.90      6745
         1.0       0.75      0.65      0.69      2300

    accuracy                           0.85      9045
   macro avg       0.82      0.79      0.80      9045
weighted avg       0.85      0.85      0.85      9045



#### SVM

In [22]:
from sklearn.svm import SVC

In [23]:
# bangun model
model = SVC(C=10, gamma=0.01, kernel='rbf')

# fit data training
model.fit(X_train, y_train)

# test model
y_pred = model.predict(X_test)

In [24]:
# print akurasi
print(f"Akurasi: {accuracy_score(y_test, y_pred):.4f}")

# print precision, recall & F-1 Score
print(classification_report(y_pred=y_pred, 
                            y_true=y_test))

Akurasi: 0.8533
              precision    recall  f1-score   support

         0.0       0.87      0.94      0.91      6745
         1.0       0.77      0.60      0.67      2300

    accuracy                           0.85      9045
   macro avg       0.82      0.77      0.79      9045
weighted avg       0.85      0.85      0.85      9045



### Using Data Testing

In [25]:
# import dataset
train = pd.read_csv('../output/adult-training.csv')
test = pd.read_csv('../output/adult-test.csv')

In [26]:
# definisi data train
X_train = train.drop(columns=['fnlwgt', 'salary'])
y_train = train.salary

In [27]:
# definisi data test
X_test = test.drop(columns=['fnlwgt', 'salary'])
y_test = test.salary

#### KNN

In [28]:
# bangun model
model = KNeighborsClassifier(n_neighbors=23, 
                             weights='distance', 
                             metric='manhattan')

# fit data training
model.fit(X_train, y_train)

# test model
y_pred = model.predict(X_test)

In [29]:
# print akurasi
print(f"Akurasi: {accuracy_score(y_test, y_pred):.4f}")

# print precision, recall & F-1 Score
print(classification_report(y_pred=y_pred, 
                            y_true=y_test))

Akurasi: 0.8536
              precision    recall  f1-score   support

         0.0       0.89      0.92      0.90     11360
         1.0       0.73      0.64      0.68      3700

    accuracy                           0.85     15060
   macro avg       0.81      0.78      0.79     15060
weighted avg       0.85      0.85      0.85     15060



#### SVM

In [30]:
# bangun model
model = SVC(C=10, gamma=0.01, kernel='rbf')

# fit data training
model.fit(X_train, y_train)

# test model
y_pred = model.predict(X_test)

In [31]:
# print akurasi
print(f"Akurasi: {accuracy_score(y_test, y_pred):.4f}")

# print precision, recall & F-1 Score
print(classification_report(y_pred=y_pred, 
                            y_true=y_test))

Akurasi: 0.8562
              precision    recall  f1-score   support

         0.0       0.88      0.94      0.91     11360
         1.0       0.77      0.59      0.67      3700

    accuracy                           0.86     15060
   macro avg       0.82      0.77      0.79     15060
weighted avg       0.85      0.86      0.85     15060



## Kesimpulan

Telah dilakukan pembangunan model menggunakan 2 teknik yang berbeda yaitu fungsi `train_test_split` dan menggunakan `adult-test.csv`. Pada kedua teknik dibangun model KNN dan SVM.

Detail model KNN:
- n = 23
- weight = distance
- metric = manhattan

Detail model SVM
- C = 10
- gamma = 0.01
- kernel = rbf



Pada `train_test_split`, akurasi KNN: 0.8547

Pada `train_test_split`, akurasi SVM: 0.8533


Pada `using data testing`, akurasi KNN: 0.8536

Pada `using data testing`, akurasi SVM: 0.8562

## Referensi

- https://medium.com/@erikgreenj/k-neighbors-classifier-with-gridsearchcv-basics-3c445ddeb657